### PR TITLE
fix: run blocking call in threadpool

### PR DESCRIPTION
**Pull request by kumaraditya303**

### CODE REVIEW

File: playwright/_impl/_browser.py [hunk 1]
```diff
@@ -106,7 +108,9 @@ async def new_context(
         storageState: Union[StorageState, str, Path] = None,
     ) -> BrowserContext:
         params = locals_to_params(locals())
-        normalize_context_params(self._connection._is_sync, params)
+        await run_in_threadpool(
```
Comment (pavelfeldman) [inline playwright/_impl/_browser.py:111]: The author of normalize_context_params might be surprised with the fact it is called on another thread. You should make `normalize_context_params` async instead and use the executor call from inside.

File: playwright/_impl/_browser_context.py [hunk 2]
```diff
@@ -214,8 +215,9 @@ async def add_init_script(
         self, script: str = None, path: Union[str, Path] = None
     ) -> None:
         if path:
-            with open(path, "r") as file:
-                script = file.read()
+            file = await run_in_threadpool(lambda: open(path, mode="r"))
```
Comment (pavelfeldman) [inline playwright/_impl/_browser_context.py:218]: It looks like you have two repetitive patterns worth extracting: `async read_file` and `async write_file`.

File: playwright/_impl/_browser_context.py [hunk 3]
```diff
@@ -260,10 +262,7 @@ async def unroute(
             )
 
     def expect_event(
-        self,
-        event: str,
-        predicate: Callable = None,
-        timeout: float = None,
+        self, event: str, predicate: Callable = None, timeout: float = None
```
Comment (pavelfeldman) [inline playwright/_impl/_browser_context.py:265]: Never change the formatter settings when changing code behavior, it adds a lot of stray changes to the code.

File: playwright/_impl/_browser_context.py [hunk 4]
```diff
@@ -298,8 +297,9 @@ async def _pause(self) -> None:
     async def storage_state(self, path: Union[str, Path] = None) -> StorageState:
         result = await self._channel.send_return_as_dict("storageState")
         if path:
-            with open(path, "w") as f:
-                json.dump(result, f)
+            file = await run_in_threadpool(lambda: open(path, mode="w"))
```
Comment (pavelfeldman) [inline playwright/_impl/_browser_context.py:300]: `await helper.write_file(path, json.dump(result))`

File: playwright/_impl/_browser_context.py [hunk 5]
```diff
@@ -310,9 +310,7 @@ async def wait_for_event(
         return await event_info
 
     def expect_page(
-        self,
-        predicate: Callable[[Page], bool] = None,
-        timeout: float = None,
+        self, predicate: Callable[[Page], bool] = None, timeout: float = None
```
Comment (pavelfeldman) [inline playwright/_impl/_browser_context.py:313]: format

File: playwright/_impl/_browser_type.py [hunk 6]
```diff
@@ -135,7 +136,9 @@ async def launch_persistent_context(
     ) -> BrowserContext:
         userDataDir = str(Path(userDataDir))
         params = locals_to_params(locals())
-        normalize_context_params(self._connection._is_sync, params)
+        await run_in_threadpool(
+            lambda: normalize_context_params(self._connection._is_sync, params)
```
Comment (pavelfeldman) [inline playwright/_impl/_browser_type.py:140]: ditto

File: playwright/_impl/_element_handle.py [hunk 7]
```diff
@@ -179,7 +180,9 @@ async def set_input_files(
         noWaitAfter: bool = None,
     ) -> None:
         params = locals_to_params(locals())
-        params["files"] = normalize_file_payloads(files)
+        params["files"] = await run_in_threadpool(
+            lambda: normalize_file_payloads(files)
```
Comment (pavelfeldman) [inline playwright/_impl/_element_handle.py:184]: ditto, also make the method async instead.

File: playwright/_impl/_element_handle.py [hunk 8]
```diff
@@ -259,10 +262,7 @@ async def query_selector_all(self, selector: str) -> List["ElementHandle"]:
         )
 
     async def eval_on_selector(
-        self,
-        selector: str,
-        expression: str,
-        arg: Serializable = None,
+        self, selector: str, expression: str, arg: Serializable = None
```
Comment (pavelfeldman) [inline playwright/_impl/_element_handle.py:265]: format

File: playwright/_impl/_element_handle.py [hunk 9]
```diff
@@ -276,10 +276,7 @@ async def eval_on_selector(
         )
 
     async def eval_on_selector_all(
-        self,
-        selector: str,
-        expression: str,
-        arg: Serializable = None,
+        self, selector: str, expression: str, arg: Serializable = None
```
Comment (pavelfeldman) [inline playwright/_impl/_element_handle.py:279]: format

File: playwright/_impl/_frame.py [hunk 10]
```diff
@@ -76,10 +77,7 @@ def __init__(
             "loadstate",
             lambda params: self._on_load_state(params.get("add"), params.get("remove")),
         )
-        self._channel.on(
-            "navigated",
-            lambda params: self._on_frame_navigated(params),
-        )
+        self._channel.on("navigated", lambda params: self._on_frame_navigated(params))
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:80]: format

File: playwright/_impl/_frame.py [hunk 11]
```diff
@@ -161,9 +159,7 @@ def predicate(event: Any) -> bool:
             return not matcher or matcher.matches(event["url"])
 
         wait_helper.wait_for_event(
-            self._event_emitter,
-            "navigated",
-            predicate=predicate,
+            self._event_emitter, "navigated", predicate=predicate
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:162]: format

File: playwright/_impl/_frame.py [hunk 12]
```diff
@@ -182,10 +178,7 @@ async def continuation() -> Optional[Response]:
         return EventContextManagerImpl(asyncio.create_task(continuation()))
 
     async def wait_for_url(
-        self,
-        url: URLMatch,
-        wait_until: DocumentLoadState = None,
-        timeout: float = None,
+        self, url: URLMatch, wait_until: DocumentLoadState = None, timeout: float = None
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:181]: format

File: playwright/_impl/_frame.py [hunk 13]
```diff
@@ -218,10 +211,7 @@ async def evaluate(self, expression: str, arg: Serializable = None) -> Any:
         return parse_result(
             await self._channel.send(
                 "evaluateExpression",
-                dict(
-                    expression=expression,
-                    arg=serialize_argument(arg),
-                ),
+                dict(expression=expression, arg=serialize_argument(arg)),
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:214]: format

File: playwright/_impl/_frame.py [hunk 14]
```diff
@@ -231,10 +221,7 @@ async def evaluate_handle(
         return from_channel(
             await self._channel.send(
                 "evaluateExpressionHandle",
-                dict(
-                    expression=expression,
-                    arg=serialize_argument(arg),
-                ),
+                dict(expression=expression, arg=serialize_argument(arg)),
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:224]: format

File: playwright/_impl/_frame.py [hunk 15]
```diff
@@ -288,10 +275,7 @@ async def dispatch_event(
         )
 
     async def eval_on_selector(
-        self,
-        selector: str,
-        expression: str,
-        arg: Serializable = None,
+        self, selector: str, expression: str, arg: Serializable = None
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:278]: format

File: playwright/_impl/_frame.py [hunk 16]
```diff
@@ -305,10 +289,7 @@ async def eval_on_selector(
         )
 
     async def eval_on_selector_all(
-        self,
-        selector: str,
-        expression: str,
-        arg: Serializable = None,
+        self, selector: str, expression: str, arg: Serializable = None
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:292]: format

File: playwright/_impl/_frame.py [hunk 17]
```diff
@@ -325,10 +306,7 @@ async def content(self) -> str:
         return await self._channel.send("content")
 
     async def set_content(
-        self,
-        html: str,
-        timeout: float = None,
-        waitUntil: DocumentLoadState = None,
+        self, html: str, timeout: float = None, waitUntil: DocumentLoadState = None
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:309]: format

File: playwright/_impl/_frame.py [hunk 18]
```diff
@@ -360,21 +338,29 @@ async def add_script_tag(
     ) -> ElementHandle:
         params = locals_to_params(locals())
         if path:
-            with open(path, "r") as file:
-                params["content"] = file.read() + "\n//# sourceURL=" + str(Path(path))
-                del params["path"]
+            file = await run_in_threadpool(lambda: open(path, mode="r"))
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:341]: `helper.read_file`

File: playwright/_impl/_frame.py [hunk 19]
```diff
@@ -360,21 +338,29 @@ async def add_script_tag(
...
         self, url: str = None, path: Union[str, Path] = None, content: str = None
     ) -> ElementHandle:
         params = locals_to_params(locals())
         if path:
-            with open(path, "r") as file:
-                params["content"] = (
-                    file.read() + "\n/*# sourceURL=" + str(Path(path)) + "*/"
-                )
-                del params["path"]
+            file = await run_in_threadpool(lambda: open(path, mode="r"))
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:356]: `read_file`

File: playwright/_impl/_frame.py [hunk 20]
```diff
@@ -479,7 +465,9 @@ async def set_input_files(
         noWaitAfter: bool = None,
     ) -> None:
         params = locals_to_params(locals())
-        params["files"] = normalize_file_payloads(files)
+        params["files"] = await run_in_threadpool(
+            lambda: normalize_file_payloads(files)
```
Comment (pavelfeldman) [inline playwright/_impl/_frame.py:469]: ditto

File: playwright/_impl/_page.py [hunk 21]
```diff
@@ -353,18 +354,12 @@ async def evaluate_handle(
         return await self._main_frame.evaluate_handle(expression, arg)
 
     async def eval_on_selector(
-        self,
-        selector: str,
-        expression: str,
-        arg: Serializable = None,
+        self, selector: str, expression: str, arg: Serializable = None
```
Comment (pavelfeldman) [inline playwright/_impl/_page.py:357]: here and below, format...

File: playwright/_impl/_page.py [hunk 22]
```diff
@@ -498,8 +481,9 @@ async def add_init_script(
         self, script: str = None, path: Union[str, Path] = None
     ) -> None:
         if path:
-            with open(path, "r") as file:
-                script = file.read()
+            file = await run_in_threadpool(lambda: open(path, mode="r"))
```
Comment (pavelfeldman) [inline playwright/_impl/_page.py:484]: `read_file`

File: playwright/_impl/_page.py [hunk 23]
```diff
@@ -542,8 +526,9 @@ async def screenshot(
         decoded_binary = base64.b64decode(encoded_binary)
         if path:
             make_dirs_for_file(path)
-            with open(path, "wb") as fd:
```
Comment (pavelfeldman) [inline playwright/_impl/_page.py:545]: `write_file`

File: playwright/_impl/_page.py [hunk 24]
```diff
@@ -741,8 +726,9 @@ async def pdf(
...
-                fd.write(decoded_binary)
+            file = await run_in_threadpool(lambda: open(path, mode="wb"))
```
Comment (pavelfeldman) [inline playwright/_impl/_page.py:729]: `read_file`

File: playwright/_impl/_path_utils.py [hunk 25]
```diff
@@ -22,3 +26,8 @@ def get_file_dirname() -> Path:
     module = inspect.getmodule(frame[0])
     assert module
     return Path(module.__file__).parent.absolute()
+
+
+def run_in_threadpool(func: Callable[..., T]) -> Awaitable[T]:
```
Comment (pavelfeldman) [inline playwright/_impl/_path_utils.py:31]: There is no point in wrapping one liner, just use run_in_executor in place.

File: playwright/_impl/_selectors.py [hunk 26]
```diff
@@ -35,8 +36,9 @@ async def register(
         if not script and not path:
             raise Error("Either source or path should be specified")
         if path:
-            with open(path, "r") as file:
-                script = file.read()
+            file = await run_in_threadpool(lambda: open(path, mode="r"))
```
Comment (pavelfeldman) [inline playwright/_impl/_selectors.py:39]: read_file

**Review by pavelfeldman — CHANGES_REQUESTED** [review]

File: playwright/_impl/_helper.py [hunk 27]
```diff
@@ -231,3 +232,33 @@ def make_dirs_for_file(path: Union[Path, str]) -> None:
...
+) -> None:
+    def inner() -> None:
+        with open(file, mode) as fh:
+            fh.write(data)
+
+    loop = asyncio.get_running_loop()
+    await loop.run_in_executor(None, inner)
+
+
+@overload
```
Comment (kumaraditya303) [inline playwright/_impl/_helper.py:248]: Overloads are for mypy else it complains

File: playwright/_impl/_helper.py [hunk 28]
```diff
@@ -231,3 +232,33 @@ def make_dirs_for_file(path: Union[Path, str]) -> None:
     if not os.path.isabs(path):
         path = Path.cwd() / path
     os.makedirs(os.path.dirname(path), exist_ok=True)
+
+
+async def async_writefile(
+    file: Union[str, Path], mode: str, data: Union[str, bytes]
```
Comment (pavelfeldman) [inline playwright/_impl/_helper.py:238]: You don't need the mode here, you can detect it based on the data type.

File: playwright/_impl/_helper.py [hunk 29]
```diff
@@ -231,3 +232,33 @@ def make_dirs_for_file(path: Union[Path, str]) -> None:
...
+async def async_readfile(file: Union[str, Path], mode: Literal["r"]) -> str:
+    ...
+
+
+@overload
+async def async_readfile(file: Union[str, Path], mode: Literal["rb"]) -> bytes:
+    ...
+
+
+async def async_readfile(file: Union[str, Path], mode: str) -> Union[str, bytes]:
```
Comment (pavelfeldman) [inline playwright/_impl/_helper.py:258]: split this into two methods, `async_readfile` that returns `bytes`, `async_read_text_file` that calls `async_readfile` and decodes.

File: playwright/_impl/_stream.py [hunk 30]
```diff
@@ -26,9 +27,8 @@ def __init__(
...
-            while True:
-                binary = await self._channel.send("read")
-                if not binary:
-                    break
-                file.write(base64.b64decode(binary))
+        while True:
+            binary = await self._channel.send("read")
+            if not binary:
+                break
+            await async_writefile(path, "wb", base64.b64decode(binary))
```
Comment (pavelfeldman) [inline playwright/_impl/_stream.py:34]: This looks like a bug, you re-write this file with the last chunk over and over.

File: playwright/_impl/_helper.py (hunk 28 above)
Comment (kumaraditya303) [reply playwright/_impl/_helper.py:238]: It is currently more explicit and better

File: playwright/_impl/_helper.py (hunk 29 above)
Comment (kumaraditya303) [reply playwright/_impl/_helper.py:258]: Sams answer as above

**Comment by pavelfeldman** [comment]
I'll have to close this because you are systematically ignoring my review comments and are marking them as resolved. Sorry about that.

**Pull request closed without merging**